### ISSUE TITLE
Error on fallback implementation

### ISSUE THREAD

**Issue by Nordiii** [ruipin/fvtt-lib-wrapper#2]
I am currently trying to implement this library and I am running into following problem when this module isn't active:

```
TypeError: Cannot read property 'value' of undefined
    at Function.register (shim.js:34)
    at PlayerData.registerMouseWheel (player-data.js:133)
    at patch_ruler (patch-ruler.js:18)
    at difficult-terrain.js:8
    at Function._call (foundry.js:2239)
    at Function.call (foundry.js:2224)
    at Canvas.draw (foundry.js:10911)
    at async Scene.view (foundry.js:29232)
    at async Game.initializeCanvas (foundry.js:5942)
    at async Game.setupGame (foundry.js:5824)
```

while my code looks like this:

```
    registerMouseWheel() {
        let self = this;
        libWrapper.register(MODULE_ID, 'canvas.activeLayer._onMouseWheel', function (wrapped, e) {
            if (!self.rulerArray.some(value => value.waypoints.length > 0))
                return wrapped(e);

            if (Date.now() - self.lastRegisteredMouseWheel < self.gameSettings.interval)
                return

            self.lastRegisteredMouseWheel = Date.now();
            if (e.deltaY < 0) {
                self.setTerrainMultiplier(self.gameSettings.increment)
                self.updateRuler(e);
            } else {
                self.setTerrainMultiplier(-self.gameSettings.increment)
                self.updateRuler(e);
            }
        }, 'MIXED');
```

I have no problem when the lib is activated but I would expect that this would still work, even when no priority is supported.

Full code if needed:
https://github.com/Nordiii/difficultterrain/tree/feature/libwrapper_support

Thanks for your time.

**Comment by ruipin** [comment]
Hi,

Thanks for the report.

So, what seems to be going on here is that the shim implementation is quite simplistic, and has one limitation I had forgot to document - it must be used on the actual place the function is defined.

`_onMouseWheel` is a function declared in the classes (e.g. `CanvasLayer`, `SightLayer`, etc).
However, `canvas.activeLayer` is an *instance* of one of those classes, and thus does not contain a member `_onMouseWheel`. Instead, Javascript will transparently inherit this member from the prototype automatically when you do an access e.g. `activeLayer._onMouseWheel`. 

While the library itself does support instance-specific overriding, the shim was written with class-specific overriding in mind only. This is because supporting instance-specific overrides in a simple shim would require adding quite a bit of extra code, and either way class-specific overriding is recommended whenever possible for reasons of resource usage and garbage collection.

A workaround would be to use `"canvas.activeLayer.prototype._onMouseWheel` to override the class directly, though then you're not overriding a specific instance's method, but all instances instead. I'll consider adding very simple "instance-specific" overrides support for the shim, and make sure to update the documentation to make all this clearer.

**Comment by Nordiii** [comment]
Thanks for the fast reply!
Sadly `canvas.activeLayer.constructor.prototype._onMouseWheel`, `CanvasLayer.prototype._onMouseWheel` nor `SightLayer.prototype._onMouseWheel` worked without an error.

I will test more tomorrow.

**Comment by ruipin** [comment]
Are you sure the hook works even with the library enabled? Was just doing random testing, e.g.

```libWrapper.register(MODULE_ID, 'canvas.activeLayer._onMouseWheel', function (wrapped, ...args) { console.log("yay"); return wrapped.apply(this, args) });```

And couldn't get the hook to work even with the big library. To be fair, I couldn't get it to work without libWrapper either, by doing

```canvas.activeLayer._onMouseWheel = function(e) { console.log("yay") };```

No matter what I do, Mouse Wheel keeps working fine. Could only get this to work when using `canvas._onMouseWheel`.

**Comment by Nordiii** [comment]
For `canvas.activeLayer._onMouseWheel` you would need to press ctrl/shift to fire it. And I am patching this in a `canvasReady ` hook

**Comment by ruipin** [comment]
Thanks, that was what I was missing.

I understand the issue now, and there's a corresponding bug in the library itself. It seems that instance-specific wrappers are "transparently broken" when using the library (the wrappers fire, but the original method doesn't) due to a bug in the instance-specific wrapping code, and the shim is just exposing it as an error since it does not replicate the same logic.

Thanks for the report - I'll look into fixing it ASAP.

**Comment by ruipin** [comment]
I've uploaded a pre-release fix: https://github.com/ruipin/fvtt-lib-wrapper/releases/tag/v1.0.4

This seems to pass all the testcases, fixes instance-wrapping, and adds official support for inherited-method wrapping (which is what you're trying to do). With this fix, the full libWrapper library should work correctly in all cases (hopefully).

In addition, I've updated the shim to give a more explicit error when the target of the wrapper isn't found, and I've updated the documentation to specify that the shim has this limitation. There is no easy way I could think of to implement inherited-method wrapping in the shim while still behaving close to the full library, so for now that is not supported.

I recommend you wrap the class directly, for example by wrapping `PlaceablesLayer.prototype._onMouseWheel` (which is the implementation that seems to be used by all the `PlaceablesLayer` children, such as the `TokenLayer`) and `Canvas.prototype._onMouseWheel` (for the global canvas).

**Comment by Nordiii** [comment]
Thanks for your quick feedback and work! I will test it later.

**Comment by Nordiii** [comment]
Ok Tested with `PlaceablesLayer.prototype._onMouseWheel` and it does indeed monkeypatch the function and my code does work, but I can't call `wrapped(e)`.
My code:
```
libWrapper.register(MODULE_ID, 'PlaceablesLayer.prototype._onMouseWheel', function (wrapped, e) {
            if (!self.rulerArray.some(value => value.waypoints.length > 0))
                return wrapped(e);   // <- Line 139

            if (Date.now() - self.lastRegisteredMouseWheel < self.gameSettings.interval)
                return;
            self.lastRegisteredMouseWheel = Date.now();
            if (e.deltaY < 0) {
                self.setTerrainMultiplier(self.gameSettings.increment)
                self.updateRuler(e);
            } else {
                self.setTerrainMultiplier(-self.gameSettings.increment)
                self.updateRuler(e);
            }
        }, 'MIXED');
``` 
The error message: 
```
Uncaught TypeError: Cannot read property 'preview' of undefined
    at _onMouseWheel (foundry.js:13262)
    at TokenLayer.<anonymous> (player-data.js:139)
    at TokenLayer.wrapper (shim.js:34)
    at KeyboardManager._onWheel (foundry.js:2497)
```

**Comment by ruipin** [comment]
Seems you are correct, I can reproduce it when not explicitly setting the `this` object on the call:



With the v1.0.4 shim, this fails:
```libWrapper.register('navbar-tweaks', 'PlaceablesLayer.prototype._onMouseWheel', function (wrapped, ...args) {
  console.log("yay");
  return wrapped(...args);
});```

And this works:
```libWrapper.register('navbar-tweaks', 'PlaceablesLayer.prototype._onMouseWheel', function (wrapped, ...args) {
  console.log("yay");
  return wrapped.apply(this, args);
});```

I'll have to figure out what is going on here. Good news is that both work with the v1.0.4 full library.

**Comment by ruipin** [comment]
I've pushed a new v1.0.4 candidate, which should fix this.

See https://github.com/ruipin/fvtt-lib-wrapper/blob/v1.0.4/shim/shim.js

Seems I forgot to bind the `this` parameter to the original method, before passing it to the shim, and none of my shim testcases were using the `wrapped(arg1, arg2, argN)` syntax so I never detected this. Originally, I enforced the usage of `apply`/`call`, but recently removed this requirement for ease-of-use and forgot to update the shim.

**Comment by ruipin** [comment]
I've published v1.0.4 as an official release. From my point-of-view, this issue should be mostly fixed - let me know if you still see issues, otherwise I'll close it.

Unfortunately, as discussed, instance-specific/inherited-method wrapping in the shim is quite a bit more difficult than it seems. I've documented the fact the shim doesn't support this, and added an explicit `throw` to catch these situations. I've created an issue #3  describing this "feature request", for future consideration.

And last but not least, thanks for your bug report!
